add new config to batch properties calls (#299)
Yelp/dataloader-codegen: README.md

        newKey: user_id
```

### Batch Resources with `properties` paramerters

> Instead of accepting just a list of users (`user_ids`), a batch resouce could accept both a list of users (`user_ids`) and a list of properties (`properties`):

Reviewer: ```suggestion
Instead of accepting just a list of users (`user_ids`), a batch resource could accept both a list of users (`user_ids`) and a list of properties (`properties`) to fetch about that user:
```

Author: Thanks for catching the spell errors! I installed the code spell checker on VS, it will help me in the future.

```js

> const getUserInfo = (args: { user_ids: Array<number>, properties: Array<string> }): Promise<Array<UserInfo>> =>
>     fetch('/userInfo', args);

Reviewer: ```suggestion
const users = getUsers({
  ids: [1, 2, 3],
  properties: ['firstName', 'age'],
})

/**
 * e.g. users => 
 * [
 *   { firstName: 'Alice', age: 42 },
 *   { firstName: 'Bob', age: 70 }
 * ]
 */
```

Just to clarify with example query and importantly, response

Author: I keep the line 155 and 156, I think the definition of the resource is also the part of this example.

Reviewer: sounds good

```

To batch up calls to this resouce with different `properties` for different `user_ids`, we specify `propertyBatchKey` in the config to describe the "properties" argument.
We specify `responseKey` in the config as the key in the response objects corresponds to `batchKey`.

The config for our `getUserInfoV2` would look like this:

```yaml
resources:
    getUserInfo:
        isBatchResource: true
        batchKey: user_ids
        newKey: user_id
        propertyBatchKey: properties
        responseKey: id
```

**IMPORTANT NOTE**
To use this feature, there are several restrictions. (Please open an issue if you're interested in helping us support other use cases):

1. The resouce accept two non-optional `list` type parameters, like this:

```
ids: str[] (should be the batchKey)
properties: str[] (should be the propertyBatchKey)
```

2. `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))

> 3. All `properties` should be optional in the response object. To enfore that, we require `swaggerLink`, `swaggerPath` and `httpMethod` defined in the config.

Reviewer: ```suggestion
3. All `properties` MUST be optional in the response object. To enfore that, we require `swaggerLink`, `swaggerPath` and `httpMethod` defined in the config.
```

https://www.ietf.org/rfc/rfc2119.txt

Reviewer: per https://github.com/Yelp/dataloader-codegen/pull/299/files#r675871585

I'd change "The flow types currently don't handle the nullability of these properties correctly, so to enforce this, we recommend a build step to ensure that the underlying types are always set as maybe types.

4. The resouce must have a one-to-one correspondence between the input "properties" and the output "properties".

>    (i.e. if we request property "name", the response must have "name" in it, and no extra data assciated with it.)

Reviewer: nit: https://www.merriam-webster.com/words-at-play/ie-vs-eg-abbreviation-meaning-usage-difference

```suggestion
   - (e.g. if we request property "name", the response must have "name" in it, and no extra data associated with it.)
```

        newKey: user_id
```

> ### Batch Resources with `properties` paramerters

Reviewer: Can we move this to API_DOCS.md? Seems more appropriate than the main readme

Author: The reason I added it here is, we have `## Batch Resources` in README and `Batch Resources with properties paramerters`  is a feature on the top of batch resources, they are somewhat related. While, API_DOCS is more about how to write a config file and the documentation of each config. 

I think it's okay to leave it in README, but if you think it will make the README too wordy or maybe we don't want to advertise this feature, I could move it to API_DCOS. Please let me know.

Reviewer: either API_DOCS.md or could be a seperate file even - this this feels like an important but "power user" feature that doesn't need to be on the homepage

Instead of accepting just a list of users (`user_ids`), a batch resource could accept both a list of users (`user_ids`) and a list of properties (`properties`) to fetch about that user:

```js
const getUserInfo = (args: { user_ids: Array<number>, properties: Array<string> }): Promise<Array<UserInfo>> =>
    fetch('/userInfo', args);

const users = getUserInfo({
  user_ids: [1, 2, 3],
  properties: ['firstName', 'age']
});

e.g. users =>
[
  { id: 1, firstName: 'Alice', age: 42 },
  { id: 2, firstName: 'Bob', age: 70 },
  { id: 3, firstName: 'Carol', age: 50 },

> ]

Reviewer: should be surrounded by comments to avoid looking weird w/ syntax highlighting

```

To batch up calls to this resource with different `properties` for different `user_ids`, we specify `propertyBatchKey` in the config to describe the "properties" argument.
We specify `responseKey` in the config as the key in the response objects corresponds to `batchKey`.

The config for our `getUserInfoV2` would look like this:

```yaml
resources:
    getUserInfo:
        isBatchResource: true
        batchKey: user_ids
        newKey: user_id
        propertyBatchKey: properties
        responseKey: id
```

**IMPORTANT NOTE**
To use this feature, there are several restrictions. (Please open an issue if you're interested in helping us support other use cases):

> 1. The resource accept two non-optional `list` type parameters, like this:

Reviewer: ```suggestion
**Contract**

1. The resource accepts two non-optional `list` type parameters:
```

> ```

Reviewer: You can preview what this looks like here: https://github.com/Yelp/dataloader-codegen/blob/344462dc8d534f5f3510b6cff4e46bf87bd8784f/README.md#batch-resources-with-properties-paramerters (or there's vscode plugins to preview the rendered markdown live)

this probably should be indented inline with the above bullet

ids: str[] (should be the batchKey)

> properties: str[] (should be the propertyBatchKey)

Reviewer: This feels slightly ambiguous - does the function accept two arguments, the first id and the second properties? or an object with these properties? (I know it's the second...I think?) probably worth clarifying tho

Reviewer: maybe something similar to how we document other fns: https://github.com/Yelp/dataloader-codegen/blob/344462dc8d534f5f3510b6cff4e46bf87bd8784f/API_DOCS.md#getloaders-arguments

```js
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>

Author: In the swagger spec, the endpoint takes 2 arguments. But if we are talking about the dataloader arguments, it takes an object like this: 
```
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>
```
I think we want restrictions on swagger spec here. Any more thoughts?

Reviewer: Not sure I understand the question here?

Author: My question: whether we should put (the argument definition from the swagger spec)
```
ids: str[] (should be the batchKey)
properties: str[] (should be the propertyBatchKey)
``` 
vs what you have here 
```
({
  // this is the batchKey
  ids: Array<string>,
  // this is the propertyBatchKey
  properties: Array<string>,
}): Array<T>
```

Sorry, I am little confused what's `the function` and `the resources` you are referring to.

Author: nvm, I think I get what you mean.

```

> 2. `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))

Reviewer: ```suggestion
2. In the response, `properties` are spread at the same level as the `responseKey`. (see `getFilmsV2` in [swapi example](./examples/swapi/swapi-server.js))
```
